### ISSUE TITLE
Is 'git lfs pull' supposed to work like 'git pull', but more optimized for LFS?

### ISSUE THREAD

**Issue by TTimo** [git-lfs/git-lfs#2594]
Is 'git lfs pull' supposed to work like 'git pull', but more optimized for LFS?
.. or am I misunderstanding what the LFS commands do?

When I need to update from my remote, I type 'git lfs pull' .. nothing happens, then I type 'git pull' and content starts to download. Except my understanding is that 'git pull' will be requesting files one by one which is significantly slower than what a more git aware operation could do?

What's the proper way? I've seen discussions about 'git lfs clone' being faster and better than 'git clone' .. is there no advantage to 'git lfs pull'? am I not using it the correct way? is it bugged?

**Comment by technoweenie** [comment]
Here's a quick description of the commands:

* `git lfs fetch` downloads objects to `.git/lfs/objects`
* `git lfs checkout` replaces objects in your working directory from `.git/lfs/objects`. Does _not_ download anything.
* `git lfs pull` is a combination of `git lfs fetch` and `git lfs checkout`

`git lfs clone` is the only wrapper command, which was added as a workaround for performance. It is deprecated in v2.3.0 though, as `git clone` should be about as fast as `git lfs clone` when used with Git v2.15+.

The reason why `git lfs pull` and `git lfs clone` are preferred is that they use LFS's built-in transfer queue object, which downloads files in parallel. If you just use `git clone` or `git pull`, LFS objects are replaced through Git's `smudge` filter.  This filter's implementation was very basic, and not designed for any sort of concurrency. Git v2.11 introduced a new filter protocol that no longer runs one `smudge` command per file, and Git v2.15 improves it to support concurrency better.

**Comment by TTimo** [comment]
Thanks for the information! So I guess the question becomes more specific:

Why does `git lfs fetch` do nothing on my system, when `git fetch` does actually find content to download? According to @technoweenie's comment both should be downloading content from remote, and `git lfs fetch` should be the faster, preferred option? (At least until lfs is >= v2.3.0 and git is >= v2.15)

**Comment by technoweenie** [comment]
`git fetch` pulls down new content from external git remotes.

`git lfs fetch` scans your current repository for lfs objects to download. Without any options, it looks at the tree of your current checked-out commit. So, if the command is doing nothing, then it's because it found no objects to download. Check out the [man page](https://github.com/git-lfs/git-lfs/blob/master/docs/man/git-lfs-fetch.1.ronn) to see all the available options.

**Comment by TTimo** [comment]
So then - unlike cloning, where `git lfs clone` is more efficient - there is no more efficient way to do a pull from a remote than to do `git pull`, correct?

**Comment by technoweenie** [comment]
You can disable automatic smudging, requiring a `git lfs pull` after checking out a new commit.

https://github.com/git-lfs/git-lfs/blob/5703b808220103063ededf5c6149c93201be9e07/docs/man/git-lfs-install.1.ronn#L36-L39

The `git lfs clone` wrapper was added because there's no easy way to enable this for a repository you haven't cloned yet.

**Comment by TTimo** [comment]
I tried this, and the first part works as advertised - `git pull` doesn't smudge and is fast, then `git lfs pull` does a single, fast download for the content.

However, I get a `Could not check out "filepath"` error for each of the updated files during the `git lfs pull` operation. And the files remain as pointers and not as actual content - any idea what's going on? If I run `git lfs checkout` I keep getting those errors.

```
git version 2.13.1.windows.2
git-lfs/2.3.0 (GitHub; windows amd64; go 1.8.3; git 4dd2bf73)
```

**Comment by technoweenie** [comment]
Doh, looks like the details of an error aren't being reported. Can you try this build? Just unzip somewhere and run something like `path/to/custom/git-lfs checkout`.

[git-lfs-windows-amd64-2.3.0.zip](https://github.com/git-lfs/git-lfs/files/1315514/git-lfs-windows-amd64-2.3.0.zip)

```diff
diff --git a/commands/pull.go b/commands/pull.go
index a825af5f..6125dae2 100644
--- a/commands/pull.go
+++ b/commands/pull.go
@@ -64,7 +64,7 @@ func (c *singleCheckout) Run(p *lfs.WrappedPointer) {
                        // acceptable error, data not local (fetch not run or include/exclude)
                        LoggedError(err, "Skipped checkout for %q, content not local. Use fetch to download.", p.Name)
                } else {
-                       FullError(fmt.Errorf("Could not check out %q", p.Name))
+                       FullError(fmt.Errorf("Could not check out %q: %s", p.Name, err))
                }
                return
        }
```

**Comment by TTimo** [comment]
This is the error:

`Could not check out "filepath": Could not create working directory file: open filepath: Access is denied.`

These files are checked in readonly. If I manual removing 'Read-only' from their properties tab the operation succeeds.

**Comment by technoweenie** [comment]
Ahh, yeah that would explain it. Unfortunately, I don't think LFS is going to work with readonly files. We'd need deeper integration with Git I think.

**Comment by TTimo** [comment]
Doesn't it just need to delete or make the text pointer file writable before getting the real content? That's what I did manually after all.

Anyway. that would be a separate issue. The question of how to do a faster pull by breaking into two steps is well adressed and documented here now. Thanks for all the info and the follow ups!
